cliquenet/src/lib.rs

```diff
@@ -391,22 +422,33 @@ where
                     }
                 },
                 // One of our connection attempts completed.
-                Some(tt) = self.connect_tasks.join_next() => match tt {
-                    Ok((s, t)) => {
-                        let Some(k) = self.lookup_peer(&t) else {
-                            continue
-                        };
-                        // We only keep the connection if our key is larger than the remote,
-                        // or if we do not have a connection for that key at the moment.
-                        if k < self.key || !self.active.contains_key(&k) {
-                            self.spawn_io(k, s, t)
-                        } else {
-                            warn!(node = %self.key, peer = %k, "dropping new connection");
+                Some(tt) = self.connect_tasks.join_next_with_id() => {
+                    match tt {
+                        Ok((id, (s, t))) => {
+                            let Some(k) = self.lookup_peer(&t) else {
+                                warn!(
+                                    node = %self.key,
+                                    peer = ?t.get_remote_static().and_then(|k| x25519::PublicKey::try_from(k).ok()),
+                                    addr = ?s.peer_addr().ok(),
+                                    "connected to unknown peer"
+                                );
+                                self.task2key.remove(&id);
+                                continue
+                            };
+                            self.on_connect_task_end(id);
+                            // We only keep the connection if our key is larger than the remote,
+                            // or if we do not have a connection for that key at the moment.
+                            if k < self.key || !self.active.contains_key(&k) {
+                                self.spawn_io(k, s, t)
+                            } else {
+                                debug!(node = %self.key, peer = %k, "dropping new connection");
+                            }
```

Comment:
how do we know that we need to remove only from `task2key` and not also from `connecting`?
Seems simpler to just do:
```suggestion
                            self.on_connect_task_end(id);
                            if let Some(k) = self.lookup_peer(&t) {
                                // We only keep the connection if our key is larger than the remote,
                                // or if we do not have a connection for that key at the moment.
                                if k < self.key || !self.active.contains_key(&k) {
                                    self.spawn_io(k, s, t)
                                } else {
                                    debug!(node = %self.key, peer = %k, "dropping new connection");
                                }
                            }
```

Reply:
I simplified that section.